controllers/components/kserve/kserve_controller.go

```diff
@@ -19,40 +19,94 @@ package kserve
 import (
 	"context"
 
-	"k8s.io/apimachinery/pkg/runtime"
+	templatev1 "github.com/openshift/api/template/v1"
+	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	admissionregistrationv1 "k8s.io/api/admissionregistration/v1"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	networkingv1 "k8s.io/api/networking/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	extv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
 	ctrl "sigs.k8s.io/controller-runtime"
-	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
 
 	componentsv1 "github.com/opendatahub-io/opendatahub-operator/v2/apis/components/v1"
+	featuresv1 "github.com/opendatahub-io/opendatahub-operator/v2/apis/features/v1"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/controller/actions/deploy"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/controller/actions/render/kustomize"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/controller/actions/security"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/controller/actions/updatestatus"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/controller/predicates/resources"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/controller/reconciler"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/metadata/labels"
 )
 
-// KserveReconciler reconciles a Kserve object.
-type KserveReconciler struct {
-	client.Client
-	Scheme *runtime.Scheme
-}
+// NewComponentReconciler creates a ComponentReconciler for the Dashboard API.
+func NewComponentReconciler(ctx context.Context, mgr ctrl.Manager) error {
+	_, err := reconciler.ComponentReconcilerFor(mgr, componentsv1.KserveInstanceName, &componentsv1.Kserve{}).
+		// operands - owned
+		Owns(&corev1.Secret{}).
+		Owns(&corev1.Service{}).
+		Owns(&corev1.ConfigMap{}).
+		Owns(&corev1.ServiceAccount{}).
+		Owns(&rbacv1.Role{}).
+		Owns(&rbacv1.RoleBinding{}).
+		Owns(&rbacv1.ClusterRole{}, builder.WithPredicates(clusterRolePredicates)).
+		Owns(&rbacv1.ClusterRoleBinding{}).
+		Owns(&templatev1.Template{}, builder.WithPredicates(compareHashPredicate)).
```

Comment:
The `ovms` template seems to sometimes get a resourceVersion change which triggers a new reconcile, so the predicate here just proves that nothing we care about has changed, and therefore no need to reconcile again for it. I'll put put a comment in the code about this, and about the clusterRolePredicate used on L54 above too.

Reply:
I don't think the `cache` is doing anything, every change can effectively trigger a reconciliation, I guess you were talking about a specific predicate that usually checks for the generation `changes`, right ?

Reply:
i was thinking two things:
1.  here is to not trigger reconcile by using predicate: when template resource is changed a on `resourceVersion` value. (i cannot think of why this could happen but it is a different question so to understand it)
2. if we do not have this predicate, then it comes to normal `cache` part, and cache wont render resource because it exclude changes on `resourceVersion`, right

Reply:
1. for some resource especially the Open Shift ones, every time you perform an action on a resource, then the `resourceVersion` changes, regardless of if the resource has actually been changed
2. the `render` cache won't trigger the re-rendering of the manifests, but the `deploy` action will deploy it as the `resourceVersion` has changed, which itself causes the `resourceVersion` to be updated, hence a loop :)